core/src/main/java/com/linecorp/armeria/client/endpoint/StickyStrategy.java

```diff
@@ -0,0 +1,140 @@
+         * @param ctx The {@code ClientRequestContext} request context
+         * @return A key related to current request, which used to load balance
+         */
+        long hashOf(ClientRequestContext ctx);
+    }
+
+    /**
+     * An interface represents strategy to hash a ClientRequestContext to a long.
+     */
+    @FunctionalInterface
+    public interface EndpointHasher {
+        /**
+         * Hash an {@code Endpoint} to a Long.
+         * @param endpoint The {@code Endpoint}
+         * @return A key related to the {@code endpoint}, which used to load balance
+         */
+        long hashOf(Endpoint endpoint);
+    }
+
+    /**
+     * Return new StickyStrategy.
+     * @param requestHasher A {@code RequestContextHasher}
+     */
+    public StickyStrategy(RequestContextHasher requestHasher) {
+        this.requestHasher = requestHasher;
+        this.endpointHasher = DEFAULT_ENDPOINT_HASHER;
+    }
+
+    /**
+     * Return new StickyStrategy.
+     * @param requestHasher A {@code RequestContextHasher}
+     * @param endpointHasher A {@code EndpointHasher}
+     */
+    public StickyStrategy(RequestContextHasher requestHasher,
+                          EndpointHasher endpointHasher) {
+        this.requestHasher = requestHasher;
+        this.endpointHasher = endpointHasher;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new StickySelector(requestHasher, endpointGroup, this);
+    }
+
+    private static final class StickySelector implements EndpointSelector {
+
+        private final RequestContextHasher hasher;
+        private final EndpointGroup endpointGroup;
+        private final EndpointSelectionStrategy endpointSelectionStrategy;
+        private static final TreeMap<Long, Endpoint> buckets = new TreeMap<>();
+
+        StickySelector(RequestContextHasher hasher, EndpointGroup endpointGroup,
+                       EndpointSelectionStrategy endpointSelectionStrategy) {
+            this.hasher = hasher;
+            this.endpointGroup = endpointGroup;
+            this.endpointSelectionStrategy = endpointSelectionStrategy;
+            if (endpointGroup instanceof DynamicEndpointGroup) {
+                DynamicEndpointGroup dynamicEndpointGroup = (DynamicEndpointGroup) endpointGroup;
+                dynamicEndpointGroup.addListener(this::rebuildBuckets);
+            }
+
+            // build buckets for first time
+            rebuildBuckets(endpointGroup.endpoints());
+        }
+
+        private synchronized void rebuildBuckets(List<Endpoint> endpoints) {
+            //TODO(huydx): is there any better way to rebuild bucket tree without clear it all?
+            // we may need a kind of "change set" of endpoints to explicitly delete/add
+            buckets.clear();
+            endpoints.forEach(endpoint -> buckets.put((long)endpoint.hashCode(), endpoint));
```

Comment:
- Probably better declaring `buckets` as a `volatile` field and create a new `TreeMap` on every `rebuildBuckets`?